### ISSUE TITLE
Support multiple ports in transparent mode

### ISSUE THREAD

**Issue by Prinzhorn** [mitmproxy/mitmproxy#4656]
#### Problem Description

I always come up with the most esoteric use-cases, but here we go. I'm pentesting an FTP setup and put mitmproxy in front of the command port (usually 21). This way I can mess with the commands traffic. However, due to the way FTP works this breaks the protocol, because the data connection (in passive mode through some random port) is not proxied through mitmproxy and hence comes from a different client and is of course blocked.

#### Proposal

I'd love to proxy everything in a given range, e.g. 30000 to 30009. With Pure-FTPd I can specify from which range it will pick the random port for each data connection. This would require

1. Creating multiple listeners on different ports
2. Mapping each of them to the reverse destination

#### Alternatives

Launching multiple mitmproxy instances, but that's no fun and I don't have everything in one UI.

#### Additional context

Burp (which is limited to HTTP) allows launching multiple listeners and displays the port in the flow list so you know which listener the traffic went through.

**Comment by mhils** [comment]
Yup! It would be really cool if we could have mitmproxy listen on port 8080 as an HTTP proxy, port 1080 as a SOCKS proxy, user-configurable ports for custom reverse proxy destinations, etc. by default.

**Comment by stellarpower** [comment]
I may be on the wrong thread here as it's been years since I've used mitmproxy, but I arrived here trying to find a solution for my use-case, and if this issue fits the context of tihs issue then it seems to me to be a bit more of a pedestrian example.

I have a pretty simple Podman setup for a number of app servers, and an nginx-based reverse proxy. The reverse proxy handles all the TLS and some basic security and just forwards different URL paths to different app servers. these servers are different versions of the same program - we want several up at once to test before making changes.

My aim is just to see all requests coming in and out of the app servers. So this includes intercepting between nginx and the app server for the Gateway<--->App server reverse proxy, and also setting all the app servers to use an mitmproxy instance as their explicit proxy, as the app servers call back out to some APIs over the Internet:

![Diagram](https://user-images.githubusercontent.com/5004545/125620516-1624f831-9141-4557-82aa-706e938b7fae.png)

Please exclude the somewhat crappy drawing. I've drawn the red dashed box around what I would like to capture. We can't really put mitmproxy in front of the reverse proxy as the dummy certificate would cause problems. 

I believe setting the (multiple) app servers to forward their outgoing requests to mitmproxy is the easy part, and I should be able to see requests out of all the app servers form the same mitmweb interface. However, I'd now also like to intercept the requests going into the app servers form the reverse proxy, without having to set up multiple mitmproxy instances. nginx just uses a simple proxy_pass to the app server, so if it would be possible to set a header in the nginx configs that represents where mitmproxy should pass the request, and instead proxy_pass to mitmproxy, then this owuld be a nice setup for me, involving at most two mitmproxy instances. Currently the workaround (if I'm not just missing an existing easy way to accomplish this) would be to add a second reverse proxy, so mitmproxy would sit inbetween the gateway and that. The gateway would just decrypt and forward the lot through mitmproxy to the second reverse proxy. which would then perform the routing to the app servers.

I would have no problem with multiple instances of mitmproxy, except that it would quickly grow out of control and require complex routing to ensure each app server was speaking to the right one. I think if this oculd be wrapped internally, so that as mentioned, the mitmproxy/mitmweb interfaces grouped in one output the details from several subinstances inside, then that might be a nice option. There's no reason why anything else owuld need to be shared between the subinstances at a granular level for me, it's just having a workable setup, ideally, one mitmproxy that can spawn up two handlers for each app server internally.

**Comment by Prinzhorn** [comment]
> Yup! Taking this one step further: It would be really cool if we could have mitmproxy listen on port 8080 as an HTTP proxy, port 1080 as a SOCKS proxy, user-configurable ports for custom reverse proxy destinations, etc. by default.

Correct me if I'm wrong, but this now works on main, right? Or is any of the things you've listed limited?
